File: Day 024/Problem 725. Split Linked List in Parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        vector<ListNode*> result;

        int listLength = 0;
        ListNode* ptr = head;
        while (ptr) {
            listLength++;
            ptr = ptr->next;
        }

        int quotient = listLength / k;
        int remainder = listLength % k;

        ptr = head;
        ListNode* list = ptr;

        while (ptr) {
            int size = quotient;
            if (remainder > 0) {
                size++;
                remainder--;
            }

            for (int i = 0; i < size - 1; i++) {
                ptr = ptr->next;
            }

            ListNode* next = ptr->next;
            ptr->next = nullptr;
            ptr = next; 

            result.push_back(list);
            list = ptr;
        } 

        while (result.size() < k) {
            result.push_back(nullptr);
        }

        return result;
    }
};

File: Day 090/Problem 725. Split Linked List in Parts.java
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) {this.val = val;}
    ListNode(int val, ListNode next) {this.val = val; this.next = next;}
}

class Solution {
    private int getLength(ListNode head) {
        int result = 0;
        for (ListNode ptr = head; ptr != null; ptr = ptr.next) {
            result++;
        }

        return result;
    }

    public ListNode[] splitListToParts(ListNode head, int k) {
        ListNode[] parts = new ListNode[k];
        int totalLength = getLength(head);

        int baseSize = totalLength / k;          // Minimum number of nodes in each part
        int extraNodes = totalLength % k;        // Extra nodes to distribute one per part

        ListNode current = head;
        for (int i = 0; i < k && current != null; i++) {
            parts[i] = current;

            int currentPartSize = baseSize + (extraNodes-- > 0 ? 1 : 0);
            for (int j = 1; j < currentPartSize; j++) {
                current = current.next;
            }

            // Cut off the current part and move to the next
            ListNode next = current.next;
            current.next = null;
            current = next;
        }

        return parts;
    }
}